File: erlang/06-io-operations/io_errors.erl
#!/usr/bin/env escript
%% io_errors.erl - Handling I/O errors gracefully
main(_) ->
    case file:read_file("does_not_exist.txt") of
        {ok, Bin} ->
            io:format("Read ~B bytes~n", [byte_size(Bin)]);
        {error, Reason} ->
            io:format("Could not read file: ~p~n", [Reason])
    end,
    io:format("Program continues after the error~n").

File: haskell/06-io-operations/io_errors.hs
import Control.Exception (try, IOException)

main :: IO ()
main = do
    -- try wraps the action; a failure becomes a Left, success a Right
    result <- try (readFile "does_not_exist.txt") :: IO (Either IOException String)
    case result of
        Left _    -> putStrLn "Error: could not read the file (it may not exist)."
        Right txt -> putStr txt

    putStrLn "The program continues normally after handling the error."
